tensorflow/lite/interpreter.cc

```diff
@@ -400,6 +400,13 @@ TfLiteStatus Interpreter::ModifyGraphWithDelegate(TfLiteDelegate* delegate) {
   if (status == kTfLiteDelegateError) {
     TF_LITE_ENSURE_STATUS(RemoveAllDelegates());
   }
+
+  // We already have a delegate. The default delegate in
```

Comment:
Acked this issue, and we've thought about such cases. The solution we provided is to leave this control to your app code when creating the TfLite interpreter. In short, one could use the BuiltinOpResolverWithoutDefaultDelegates op resolver defined in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/register.h to skip the application of default delegates.

Reply:
@multiverse-tf 

In tflite benchmark tool and label_image example, the flow to initialize interpreter is:
```
1. call InterpreterBuilder() to build a Interpreter.
	a. get default delegate from op_resolver
	  if (num_fp32_tensors_ > 0) {
    	    (*interpreter)->lazy_delegate_providers_ =
            op_resolver_.GetDelegates(num_threads);
  	  }

2. choose delegate from app side.
	a. iterate all delegates and pick one delegate
	b. apply ModifyGraphWithDelegate() for the chosen one

3. call interperter->AllocateTensors().
	a. call ModifyGraphWithDelegate() for "lazy_delegate_providers_".
```


In my point of view, it has flexibility to set the delegate from user/app side and it also could use the default delegate if the user doesn't set up custom delegate in step 2.

You said that we could just use BuiltinOpResolverWithoutDefaultDelegates() for builder to skip the default delegate. There will be no default delegate anymore. People should know that we should use BuiltinOpResolverWithoutDefaultDelegates() if we try to have the custom delegate like step 2.

Why could we have both(the default one and the custom delegates)?

Reply:
> @multiverse-tf
> 
> In tflite benchmark tool and label_image example, the flow to initialize interpreter is:
> 
> ```
> 1. call InterpreterBuilder() to build a Interpreter.
> 	a. get default delegate from op_resolver
> 	  if (num_fp32_tensors_ > 0) {
>     	    (*interpreter)->lazy_delegate_providers_ =
>             op_resolver_.GetDelegates(num_threads);
>   	  }
> 
> 2. choose delegate from app side.
> 	a. iterate all delegates and pick one delegate
> 	b. apply ModifyGraphWithDelegate() for the chosen one
> 
> 3. call interperter->AllocateTensors().
> 	a. call ModifyGraphWithDelegate() for "lazy_delegate_providers_".
> ```
> 
> In my point of view, it has flexibility to set the delegate from user/app side and it also could use the default delegate if the user doesn't set up custom delegate in step 2.
> 
> You said that we could just use BuiltinOpResolverWithoutDefaultDelegates() for builder to skip the default delegate. There will be no default delegate anymore. People should know that we should use BuiltinOpResolverWithoutDefaultDelegates() if we try to have the custom delegate like step 2.

Acked. There are several factors involved here:
1. The default delegate in TFLite is applied at the beginning of AllocateTensors(), which is generally called after ModifyGraphWithDelegate(..). Therefore, it's likely the default delegate is the last one to be applied after a customized user delegate. Admittedly, this is not always true. Therefore, we are reconsidering applying-delegate APIs, and you can see a similar ModifyGraphWithDelegate API in the Java binding has been marked deprecated and a different approach is preferred (i.e. via Options struct).

2. Although theoretically multiple delegates could be applied one after another in TFLite, in general, only one delegate is applied to the model graph to save certain overheads (like memory transfer overhead etc.). Also, the graph modified by the first delegate is generally not "modifiable" by a second delegate. From this perspective, if the app side choose to apply a customized delegate, it's likely the default delegate won't be able to be applied afterwards.

3. As for the benchmark model tool, I believe the default delegate will be applied last because of point 1 mentioned above. Regarding the label_image one, I need to double-check.

> 
> Why could we have both(the default one and the custom delegates)?

I think it's a great idea to have a model graph modified by the default one (esp. w/ the XNNPACK delegate) simultaneously with a graph modified by customized ones because the former could serve as a really performant fallback when the customized delegate doesn't work at runtime. However, such a feature will complicate the current implementation/logic further, and we are actively thinking about alternative approaches.

Reply:
> 3. As for the benchmark model tool, I believe the default delegate will be applied last because of point 1 mentioned above. Regarding the label_image one, I need to double-check.
> 
> > Why could we have both(the default one and the custom delegates)?
> 
> I think it's a great idea to have a model graph modified by the default one (esp. w/ the XNNPACK delegate) simultaneously with a graph modified by customized ones because the former could serve as a really performant fallback when the customized delegate doesn't work at runtime. However, such a feature will complicate the current implementation/logic further, and we are actively thinking about alternative approaches.


I haven't thought about using multiple delegates together yet. That's cool.
There problem here is that the interpreter tries to apply the default delegate(the xnnpack here) after benchmark tool/label_image uses user specific delegate.

The error messages here are confusing. It looks like there is no delegate available here.
```
# user selects xnnpack and call ModifyGraphWithDelegate()
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
INFO: Use XNNPACK acceleration.
INFO: Applied XNNPACK delegate.
# user call AllocateTensors(). And AllocateTensors() call ModifyGraphWithDelegate() with the default delegate again.
# The graph is already become immutable at this moment, so we will see the following message.
ERROR: ModifyGraphWithDelegate is disallowed when graph is immutable.
ERROR: Ignoring failed application of the default TensorFlow Lite delegate indexed at 0.
```

If we will have another api/interface for delegate setting, do you think we should use BuiltinOpResolverWithoutDefaultDelegates() instead of BuiltinOpResolver() in benchmark/label_image example?

Reply:
> > 1. As for the benchmark model tool, I believe the default delegate will be applied last because of point 1 mentioned above. Regarding the label_image one, I need to double-check.
> > 
> > > Why could we have both(the default one and the custom delegates)?
> > 
> > 
> > I think it's a great idea to have a model graph modified by the default one (esp. w/ the XNNPACK delegate) simultaneously with a graph modified by customized ones because the former could serve as a really performant fallback when the customized delegate doesn't work at runtime. However, such a feature will complicate the current implementation/logic further, and we are actively thinking about alternative approaches.
> 
> I haven't thought about using multiple delegates together yet. That's cool.
> There problem here is that the interpreter tries to apply the default delegate(the xnnpack here) after benchmark tool/label_image uses user specific delegate.
> 
> The error messages here are confusing. It looks like there is no delegate available here.
> 
> ```
> # user selects xnnpack and call ModifyGraphWithDelegate()
> INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
> INFO: Use XNNPACK acceleration.
> INFO: Applied XNNPACK delegate.
> # user call AllocateTensors(). And AllocateTensors() call ModifyGraphWithDelegate() with the default delegate again.
> # The graph is already become immutable at this moment, so we will see the following message.
> ERROR: ModifyGraphWithDelegate is disallowed when graph is immutable.
> ERROR: Ignoring failed application of the default TensorFlow Lite delegate indexed at 0.
> ```
> 

Sorry for the confusing message. Actually, we were quite aware of this and tried our best to avoid the confusion. The first error msg is the impact of after applying the 'default' xnnpack delegate (it looks like you've chosen to compile the binary w/ xnnpack delegate applied by default). The second error msg was added in https://github.com/tensorflow/tensorflow/commit/4bb7dca4465a2c2c2fc71f58187de975cff79aa1, hoping to imply "we are aware of this particular case, thus ignoring the error in applying the default delegate.".

> If we will have another api/interface for delegate setting, do you think we should use BuiltinOpResolverWithoutDefaultDelegates() instead of BuiltinOpResolver() in benchmark/label_image example?

Once we completely flip the flag to enable applying default delegate across all major platforms, we may consider doing this. But there's a strong preference to use BuiltinOpResolver in these tools so that the benefit of xnnpack delegate will be transparently enjoyed by users.

Reply:
Ok, let's ignore this error message here.